### ISSUE TITLE
Unreachable patterns are linted, but there is a chance it will be removed for conditionally-compiled enums

### ISSUE THREAD

**Issue by vityafx** [rust-lang-nursery/rust-clippy#2469]
As of [now](https://github.com/rust-lang/rust/issues/47905), the unreachable arms in the `match` statement are treated as bad style (this is good warning in general) and the warning is shown by the `rustc`. However, it can be so that the `match` arms are written so explicitly because they match a enum with conditionalliy-compiled variants. This makes this warning useless and harmful, there is a chance [that it will be removed](https://github.com/rust-lang/rust/issues/47905#issuecomment-361962356). The issue here is that the `clippy` also treats these arms as non-idiomatic way and suggest to rewrite them as `if let` statement, for example (if there are two variants in the enum and one of them is removed because of conditional compilation (feature)). I suggest to follow the issue I mentioned and try not to suggest match arm changes if they are written for enums with conditionally-compiled variants.

**Comment by Manishearth** [comment]
We can't detect that enums are conditionally compiled.

This is enough of a special case that you should `allow()` the lint. Lints aren't universal.

**Comment by vityafx** [comment]
@Manishearth afaik oli-obk told that there is a chance they would do something for detecting this.

**Comment by vityafx** [comment]
@Manishearth Also, could you tell me how can I allow them without using `clippy` as a compiler plugin, but using command-line version of it? I tried this `clippy.toml` but that did not work:

```toml
single_match = "allow"
```

**Comment by oli-obk** [comment]
@vityafx there's nothing on the clippy side that we can do. You'll need to change it in the Rust compiler so that cfg-ed code still shows up, but is ignored by the rest of the compiler. I do not really see the need for that, I just meant that there are possible avenues. It will add complexity to the compiler, and that needs to be warranted with a good enough gain. cfg-heavy code is very similar to generated or macro-expanded code. Maybe rustc could simply mark code touched by cfgs as having expansion info?

**Comment by vityafx** [comment]
This is really very sad. Don't you realize that this is really useful - instead of disabling warning making it be just correct? Because now it's not since it is emitted for conditionally-compiled code. The correct solution would still be to touch the compiler internals instead of silencing such warnings. As for complexity of the compiler - everything makes it more complex from day to day. Don't really see here a problem, except that it is not quite easy to implement this, - anyway, this is not the case it should not be done, in my opinion. If I were the rustc developer I'd do that. I'd do that even if I were not, because I want it to be the best compiler as it could be. You may just point me to the code that I need to touch or at least to point me to the places where I should look at, I may try doing this in my fork, as I can.

**Comment by oli-obk** [comment]
> Don't really see here a problem, except that it is not quite easy to implement this,

I don't think it's that complex. Basically conditional compilation would just cause its outer item to get some expansion info.

> If I were the rustc developer I'd do that. I'd do that even if I were not, because I want it to be the best compiler as it could be. You may just point me to the code that I need to touch or at least to point me to the places where I should look at, I may try doing this in my fork, as I can.

Great! Skimming through the code it looks to me like https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser.rs#L5823-L5842 and code around all calls to the `configure` method (within the `libsyntax` crate) need to add expansion info to their spans. So instead of simply filtering out the unconfigured items, you'll need to remember that any item was ignored and then inform its surrounding node about that.

**Comment by vityafx** [comment]
@oli-obk Thanks much, great info! I'll try doing my best.
